File: examples/GDSReader/GDSReader2D.cpp
#include <psGDSReader.hpp>

#include <lsToSurfaceMesh.hpp>
#include <lsVTKWriter.hpp>

namespace ps = viennaps;
namespace ls = viennals;

int main(int argc, char **argv) {
  using NumericType = double;
  constexpr int D = 2;

  // read GDS mask file
  ps::Logger::setLogLevel(ps::LogLevel::DEBUG);

  constexpr NumericType gridDelta = 0.01;
  constexpr NumericType exposureDelta = 0.005;
  double forwardSigma = 5.;
  double backsSigma = 50.;

  ls::BoundaryConditionEnum boundaryConds[D] = {
      ls::BoundaryConditionEnum::REFLECTIVE_BOUNDARY,
      ls::BoundaryConditionEnum::REFLECTIVE_BOUNDARY};

  auto mask = ps::SmartPointer<ps::GDSGeometry<NumericType, D>>::New(
      gridDelta, boundaryConds);
  mask->addBlur({forwardSigma, backsSigma}, // Gaussian sigmas
                {0.8, 0.2},                 // Weights
                0.5,                        // Threshold
                exposureDelta);             // Exposure grid delta
  ps::GDSReader<NumericType, D>(mask, "myTest.gds").apply();

  auto maskLayer = mask->layerToLevelSet(0, false);
  auto mesh = ls::SmartPointer<ls::Mesh<NumericType>>::New();
  ls::ToSurfaceMesh<NumericType, D>(maskLayer, mesh).apply();
  ls::VTKWriter<NumericType>(mesh, "maskLayer.vtp").apply();

  auto blurredLayer = mask->layerToLevelSet(0, true);
  ls::ToSurfaceMesh<NumericType, D>(blurredLayer, mesh).apply();
  ls::VTKWriter<NumericType>(mesh, "blurredLayer.vtp").apply();

  return 0;
}

File: examples/GDSReader/GDSReader2D.py
import viennaps as ps

# Set log level
ps.Logger.setLogLevel(ps.LogLevel.DEBUG)

# Parameters
gridDelta = 0.01
exposureDelta = 0.005
forwardSigma = 5.0
backsSigma = 50.0

boundaryConds = [
    ps.BoundaryType.REFLECTIVE_BOUNDARY,
    ps.BoundaryType.REFLECTIVE_BOUNDARY,
]

# Create GDS geometry object
mask = ps.GDSGeometry(gridDelta, boundaryConds)

# Add blur parameters
mask.addBlur(
    sigmas=[forwardSigma, backsSigma],
    weights=[0.8, 0.2],
    threshold=0.5,
    delta=exposureDelta,
)

# Load GDS file
reader = ps.GDSReader(mask, "myTest.gds")
reader.apply()

# Export unblurred mask layer
maskLayer = mask.layerToLevelSet(0, blurLayer=False)
mesh = ps.ls.Mesh()
ps.ls.d2.ToSurfaceMesh(maskLayer, mesh).apply()
ps.ls.VTKWriter(mesh, "maskLayer.vtp").apply()

# Export blurred mask layer
blurredLayer = mask.layerToLevelSet(0, blurLayer=True)
ps.ls.d2.ToSurfaceMesh(blurredLayer, mesh).apply()
ps.ls.VTKWriter(mesh, "blurredLayer.vtp").apply()
